coral-schema/src/main/java/com/linkedin/coral/schema/avro/SchemaUtilities.java

```diff
@@ -394,36 +380,46 @@ static boolean isUnionRecordSchemaCompatible(@Nonnull Schema leftSchema, @Nonnul
     for (Schema.Field field : leftSchemaFields) {
       if (!rightSchemaFieldsMap.containsKey(field.name())) {
         // field in leftSchema is missing in rightSchema
-        LOG.error(field.name() + " is in schema " + leftSchema.getName() + ": " + leftSchema.toString(true)
-            + ", but not in schema " + rightSchema.getName() + ": " + rightSchema.toString(true));
-        return false;
+        throw new RuntimeException(
+            field.name() + " is in schema " + leftSchema.getName() + ": " + leftSchema.toString(true)
+                + ", but not in schema " + rightSchema.getName() + ": " + rightSchema.toString(true));
       }
     }
 
     for (Schema.Field field : rightSchemaFields) {
       if (!leftSchemaFieldsMap.containsKey(field.name())) {
         // field in rightSchema is missing in leftSchema
-        LOG.error(field.name() + " is in schema " + rightSchema.getName() + ": " + rightSchema.toString(true)
-            + ", but not in schema " + leftSchema.getName() + ": " + leftSchema.toString(true));
-        return false;
+        throw new RuntimeException(
+            field.name() + " is in schema " + rightSchema.getName() + ": " + rightSchema.toString(true)
+                + ", but not in schema " + leftSchema.getName() + ": " + leftSchema.toString(true));
       }
     }
 
+    List<Schema.Field> mergedSchemaFields = new ArrayList<>();
+
     for (Schema.Field leftField : leftSchemaFields) {
       Schema.Field rightField = rightSchemaFieldsMap.get(leftField.name());
-      boolean isUnionSchemaCompatible = isUnionSchemaCompatible(leftField.schema(), rightField.schema(), strictMode);
-
-      if (!isUnionSchemaCompatible) {
-        LOG.error(leftField.name() + " is not compatible with " + rightField.name() + " for LogicalUnion operator.");
-
-        return false;
+      Schema unionFieldSchema = getUnionFieldSchema(leftField.schema(), rightField.schema(), strictMode);
+      if (unionFieldSchema == null) {
+        unionFieldSchema = getUnionFieldSchema(rightField.schema(), leftField.schema(), strictMode);
+      }
+      if (unionFieldSchema == null) { // types of leftField and rightField are not compatible
```

Comment:
Because if `unionFieldSchema == null`, that means `leftField` and `rightField` are not compatible.
`getUnionFieldSchema` and `mergeUnionRecordSchema` are similar to previous 2 functions, the difference is the previous version only returns `true` or `false`, but this patch returns the union result, if two schema are not compatible, returns `null`. 
The reason of returning union result is it makes supporting the following union easier:
```
null UNION ALL int -> [null, int]
null UNION ALL [null, int] -> [null, int]
int UNION ALL [null, int] -> [null, int]
```